### PR TITLE
KPrototypes PART 2

### PR DESCRIPTION
**Pull request by oualib**
- Unit Tests still missing
 - code formatting using BLACK

### CODE REVIEW

File: verticapy/learn/memmodel.py [hunk 1]
```diff
@@ -1527,44 +1528,254 @@ def sql_from_clusters(
...
+    )
+
+    assert not (return_distance_clusters) or not (return_proba), ParameterError(
+        "Parameters 'return_distance_clusters' and 'return_proba' cannot both be set to True."
+    )
+
+    if not (is_categorical):
+        is_categorical = [True for i in range(len(X))]
+
+    for c in clusters:
```
Comment (afard) [inline verticapy/learn/memmodel.py:1717]: Shouldn't we have the same check in the predict_from_clusters_kprotypes function?

File: verticapy/learn/memmodel.py [hunk 2]
```diff
@@ -1527,44 +1528,254 @@ def sql_from_clusters(
...
+            ("is_categorical", is_categorical, [list, np.ndarray]),
+        ]
+    )
+
+    assert not (return_distance_clusters) or not (return_proba), ParameterError(
+        "Parameters 'return_distance_clusters' and 'return_proba' cannot both be set to True."
+    )
+
+    if not (is_categorical):
+        is_categorical = [True for i in range(len(X))]
```
[the remark this answers is not in the document]
Comment (afard) [reply verticapy/learn/memmodel.py:1715]: A good set of unite tests would examine these corner cases as well.

**Review by tconsigny — APPROVED** [review]

File: verticapy/learn/memmodel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/memmodel.py:1718]: The check is done L1721

File: verticapy/learn/vmodel.py [hunk 3]
```diff
@@ -2271,10 +2274,20 @@ def to_memmodel(self, **kwds):
...
-        elif self.type in ("KMeans",):
+        elif self.type in ("KMeans", "KPrototypes"):
             attributes = {"clusters": self.cluster_centers_.to_numpy(), "p": 2}
+            if self.type == "KPrototypes":
+                attributes["gamma"] = self.parameters["gamma"]
+                attributes["is_categorical"] = [
+                    ("char" in self.cluster_centers_.dtype[key].lower())
+                    for key in self.cluster_centers_.dtype
+                ]
         elif self.type in ("KPrototypes",):
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/vmodel.py:2285]: Thanks for noticing that, -> erased

File: verticapy/learn/memmodel.py [hunk 4]
```diff
@@ -1527,44 +1528,257 @@ def sql_from_clusters(
...
+    assert not (return_distance_clusters) or not (return_proba), ParameterError(
+        "Parameters 'return_distance_clusters' and 'return_proba' cannot both be set to True."
+    )
+
+    centroids = np.array(clusters)
+
+    def compute_distance_row(X):
+        result = []
+        for centroid in centroids:
+            distance_num, distance_cat = 0, 0
```
[the remark this answers is not in the document]
Comment (oualib) [reply verticapy/learn/memmodel.py:1634]: we can add this operation in the loop but it is a plus. The automatic error message generated by Python will be enough.

**Review by afard — APPROVED** [review]
